src/Nethermind/Nethermind.Blockchain/Blocks/BlockhashStore.cs

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText:2023 Demerzel Solutions Limited
+// SPDX-License-Identifier:LGPL-3.0-only
+
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+using Nethermind.Blockchain.Find;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.State;
+
+[assembly: InternalsVisibleTo("Nethermind.Blockchain.Test")]
+[assembly: InternalsVisibleTo("Nethermind.Merge.Plugin.Test")]
+namespace Nethermind.Blockchain.Blocks;
+
+public class BlockhashStore(IBlockFinder blockFinder, ISpecProvider specProvider, IWorldState worldState)
+    : IBlockhashStore
+{
+    private static readonly byte[] EmptyBytes = [0];
+
+    public void ApplyHistoryBlockHashes(BlockHeader blockHeader)
+    {
+        IReleaseSpec spec = specProvider.GetSpec(blockHeader);
+        if (!spec.IsEip2935Enabled || blockHeader.IsGenesis || blockHeader.ParentHash is null) return;
+
+        Address? eip2935Account = spec.Eip2935ContractAddress ?? Eip2935Constants.BlockHashHistoryAddress;
+        if (!worldState.AccountExists(eip2935Account)) return;
+
+        // TODO: find a better way to handle this - no need to have this check everytime
```

Comment:
add github issue to not forget?

Reply:
I think it is fine, the only other way would be reading from state, this should be faster as the header should be in memory.

Reply:
yes, i was more thinking along the lines of adding a flag like, isBlockhashState initialized. Once its set, then we wont have to do this check for every block.

Reply:
Then you need to handle reorgs, theoretically maybe might be fine when we finalize transition block.

Reply:
ahh yes, reorgs would be a issues - lets keep it like this then